src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -171,7 +147,109 @@ public void handleException(TransportException exp) {
+                publishResponseHandler.onFailure(node, e);
+                return;
+            }
+        }
+        publishClusterStateToNode(clusterState, bytes, node, timedOutWaitingForNodes, publishTimeout, publishResponseHandler, true);
+    }
+
+    private void publishClusterStateToNode(final ClusterState clusterState, BytesReference bytes,
+                                           final DiscoveryNode node, final AtomicBoolean timedOutWaitingForNodes,
+                                           final TimeValue publishTimeout,
+                                           final BlockingClusterStatePublishResponseHandler publishResponseHandler,
+                                           final boolean sendDiffs) {
+        try {
+            TransportRequestOptions options = TransportRequestOptions.options().withType(TransportRequestOptions.Type.STATE).withCompress(false);
+            // no need to put a timeout on the options here, because we want the response to eventually be received
+            // and not log an error if it arrives after the timeout
+            transportService.sendRequest(node, ACTION_NAME,
+                    new BytesTransportRequest(bytes, node.version()),
+                    options, // no need to compress, we already compressed the bytes
+
+                    new EmptyTransportResponseHandler(ThreadPool.Names.SAME) {
+
+                        @Override
+                        public void handleResponse(TransportResponse.Empty response) {
+                            if (timedOutWaitingForNodes.get()) {
+                                logger.debug("node {} responded for cluster state [{}] (took longer than [{}])", node, clusterState.version(), publishTimeout);
+                            }
+                            publishResponseHandler.onResponse(node);
+                        }
+
+                        @Override
+                        public void handleException(TransportException exp) {
+                            if (sendDiffs && exp.unwrapCause() instanceof IncompatibleClusterStateVersionException) {
+                                logger.debug("resending full cluster state to node {} reason {}", node, exp.getDetailedMessage());
+                                sendFullClusterState(clusterState, null, node, timedOutWaitingForNodes, publishTimeout, publishResponseHandler);
+                            } else {
+                                logger.debug("failed to send cluster state to {}", exp, node);
+                                publishResponseHandler.onFailure(node, exp);
+                            }
+                        }
+                    });
+        } catch (Throwable t) {
+            logger.debug("error sending cluster state to {}", t, node);
+            publishResponseHandler.onFailure(node, t);
+        }
+    }
+
+    public static BytesReference serializeFullClusterState(ClusterState clusterState, Version nodeVersion) throws IOException {
```

Comment:
Not really sure what we would test here separately. We test serialization logic for diffs and compare it to full serialization and we also test that publishing logic switches properly between full serialization and diff serialization. During ClusterStateDiffTest all aspects of diff serialization, but I am going to enhance it by sprinkling full cluster serialization into it from time to time.